src/extensions/mountainview/MVSortingView/PreloadCheck.tsx

```diff
@@ -0,0 +1,115 @@
+import { Dialog } from '@material-ui/core';
+import React, { Fragment, FunctionComponent, useContext, useEffect, useMemo, useRef, useState } from 'react';
+import { createCalculationPool, HitherContext } from '../../common/hither';
+import { Recording, Sorting } from '../../extensionInterface';
+
+interface ChildProps {
+    preloadStatus?: 'waiting' | 'running' | 'finished'
+}
+
+type Props = {
+    sorting: Sorting
+    recording: Recording
+    children: React.ReactElement<ChildProps>
+}
+
+const calculationPool = createCalculationPool({maxSimultaneous: 6})
+
+const PreloadCheck: FunctionComponent<Props> = ({ recording, sorting, children }) => {
+    const hither = useContext(HitherContext)
+    const sortingObject = sorting.sortingObject
+    const recordingObject = recording.recordingObject
+    const [error, setError] = useState<Error | null>(null)
+    const [status, setStatus] = useState<'waiting' | 'running' | 'finished'>('waiting')
+    const [message, setMessage] = useState<string>('')
+    const runningState = useRef<{sortingObject: any, recordingObject: any}>({sortingObject: sorting.sortingObject, recordingObject: recording.recordingObject})
+
+    const matchesRunningState = useMemo(() => ((x: {recordingObject: any, sortingObject: any}) => (
+        (runningState.current.sortingObject === x.sortingObject) && (runningState.current.recordingObject === x.recordingObject)
+    )), [])
+
+    useEffect(() => {
+        if (status === 'waiting') {
+            runningState.current = {recordingObject, sortingObject}
+            setStatus('running')
+            ;(async () => {
+                try {
+                    setMessage('Checking sorting local...')
+                    const result1 = await hither.createHitherJob('preload_check_sorting_downloaded', {sorting_object: sortingObject}, {useClientCache: false, calculationPool}).wait() as {isLocal: boolean}
+                    if (!matchesRunningState({recordingObject, sortingObject})) return
```

Comment:
These checks (l. 37 - 71, the whole `try` block for downloading the data) are all very similar--I think it's possible to simplify a lot by introducing a function here. (Would probably make it easier to follow too.)

Also to make sure I'm clear--the purpose of these hither calls is to make sure that the data (respectively, sortings, recordings, and snippets) have been downloaded/processed and are now available locally, right?